crypto/dh/dh_lib.c

```diff
@@ -78,14 +78,16 @@ static DH *dh_new_intern(ENGINE *engine, OSSL_LIB_CTX *libctx)
     if (ret == NULL)
         return NULL;
 
-    ret->references = 1;
     ret->lock = CRYPTO_THREAD_lock_new();
     if (ret->lock == NULL) {
         ERR_raise(ERR_LIB_DH, ERR_R_CRYPTO_LIB);
         OPENSSL_free(ret);
         return NULL;
     }
 
+    if (!CRYPTO_NEW_REF(&ret->references, 1))
+        goto err;
```

Comment:
You do not want to use goto err; here, it will call DH_free() which wants to do CRYPTO_DOWN_REF().